Run GraphQL query based on cursor position.

bwlt opened the pull request:
If there are multiple operations in the GraphQL query it automatically choose
which operation to run based on the cursor position.

bwlt committed:
Run GraphQL query based on cursor position.

If there are multiple operations in the GraphQL query it automatically choose
which operation to run based on the cursor position.

gschier commented:
This is a very well done PR, thanks! 

I will also leave some actual review comments but I just wanted to have a brief high-level discussion about this feature first.

While I do think this feature is useful (and I know it's part of GraphiQL), it seems like there is potential to cause confusion since there's no obvious indication that the cursor position is controlling the selected query.

A quick improvement could be to highlight the selected query name based on the currently active `OperationName`. I'm perfectly happy making this (or similar) improvement after merging this PR. However, if you felt like adding this yourself that would be awesome too.

Thoughts?

bwlt commented:
Ok, how to highlight? (in terms of style)
I thought on these two alternatives:

- the selected query have syntax color
![screen shot 2018-06-19 at 14 16 11](https://user-images.githubusercontent.com/4998750/41596815-cda9b91e-73cb-11e8-8bcb-5c0e528c4688.png)

- the selected query have a **bolder** operation name
![screen shot 2018-06-19 at 14 16 35](https://user-images.githubusercontent.com/4998750/41596816-cdc585ea-73cb-11e8-88a5-b0d00f9074ae.png)

(I'm not good on design stuff 😅)

gschier commented:
Ooo I like grayscaling everything else but I'm not sure how difficult that is to implement. If it's not easy, I would also be fine with using Codemirror's `markText` function to highlight the background of the query name.

Here's a code snippet of how it works.

```js
editor.markText({line: 6, ch: 26}, {line: 6, ch: 42}, {className: "styled-background"});
```

Setting up the precise colors to use may be tricky since it has to work with the Insomnia theme system. But, if you want to add the basic implementation I can help fine-tune it if needed.

gschier commented:
Docs, if needed https://codemirror.net/doc/manual.html#api_marker

bwlt committed:
Highlight current operation

bwlt commented:
I have worked on it using codemirror textmarker.
The result is nice! Using graphql documentAST helped a lot.
Let me work on a little issue introduced in this PR; a little refactor is
necessary :)

Il mer 20 giu 2018, 05:30 Gregory Schier <notifications@github.com> ha
scritto:

> Docs, if needed https://codemirror.net/doc/manual.html#api_marker
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/getinsomnia/insomnia/pull/990#issuecomment-398613180>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AExGXuAYCx8cMpw--yKfhVwQsqHb0nKFks5t-cHMgaJpZM4Uqx8q>
> .
>

gschier commented:
Awesome! I'm looking forward to it 😄 

Let me know if you need any help! 👍

bwlt committed:
Refactor. Prevent some query change issues.

If the current operation name gets deleted the new current operation
is set as fallback to the first operation name.
This refactor includes a `_documentAST` property on GraphQLEditor component,
so it is locally available (it avoids multiple graphql `parse` function calls).
Also a _queryEditor reference is added to the component.

bwlt commented on packages/insomnia-app/app/ui/css/components/graph-ql-editor.less:
```diff
@@ -21,4 +21,10 @@
   }
+
+  &__disabled-definition {
+    &&& {
```
@gschier sorry, this is a dirty trick to increase the specificity (always better than `!important`, inspired from [styled-components](https://www.styled-components.com/docs/faqs#how-can-i-override-styles-with-higher-specificity))

gschier committed:
Merge branch 'develop' into feature/graphql-choose-operation-name

jsejcksn commented:
It seems like this change didn't make it into the docs: https://support.insomnia.rest/article/61-graphql
